core/messagechains/src/lib.rs

```diff
@@ -0,0 +1,583 @@
+use std::collections::HashMap;
+use std::collections::VecDeque;
+use std::fmt::Debug;
+use std::hash::Hash;
+
+use log;
+
+pub mod sha256;
+pub mod wasm_wrapper;
+
+/// Generic device ID
+pub trait DeviceId: Sized + Hash + Eq + Ord + Clone + Debug + AsRef<[u8]> {}
+
+/// Blanket implementation of the [`DeviceId`] type for [`String`]
+impl DeviceId for String {}
+
+/// Generic wrapper over a cryptographic message digest
+pub trait MessageDigest: Sized + Hash + Eq + Ord + Clone + Debug + Default {}
```

Comment:
Similarly, is there a case where we don't want to just use a _particular_ hash function? (and even if we do, the hash functions already have a generic interface in the `digest` crate)